### DISCUSSION TITLE
How to define precision/units/epsilon?

### DISCUSSION THREAD

**Discussion by hrj** [jscad/OpenJSCAD.org#1310] [General]
Hi,

Firstly, many thanks for this project!

I have been building a fairly complex parametric model using jscad, but the linear extrusion step some times throws the exception "geometry is not closed". The model has some points which are 1/10th of a mm away, and I think these points get erroneously merged together during the repair step.

If I scale up the model to 10x, the exception goes away, but then scaling up the model messes with the units of the SVG and STL outputs. And scaling down the model for export could introduce its own errors.

I see that there are some constants defined, such as EPS and spatialResolution. But they are in the 100s of nanometer scale. So, distances of .1 mm shouldn't be a problem?

Do you have any recommendations to solve this?

And, do you have tests with small models? I wish I could simply share my model for testing, but it's part of a larger project, and difficult to separate, but I will try to do that.

**Comment by z3dev** [comment]
@hrj Welcome!

Some good questions!

Yes. Most of the modeling library doesn't depend on precision and should be fine. However, there are times when the EPS is checked between points, etc.

There are checks when converting from 2D to 3D, and exporting. It seems that you hit one of those checks.

Some hints for converting 2D to 3D
- all 2D outlines should be fully closed, no gaps. geom2.validate() can be called anytime to check
- snap() can also be called to reset all the points to a 'normalized' EPS

Also, there are know issues with converting from SVG to 2D geometries. SVG paths can cross, touch, rotate, etc. Some SVG will not convert.

If you still have problems then please output the 2D geometry to JSON. We can then take a look at the contents, and maybe make some recommendations. JSON.stringify(object)

**Reply by hrj** [reply]
Thank you @z3dev !

The 2D outlines have been created through boolean operations in jscad itself, not by importing.

I managed to create a test case for this. In fact, it exposes two issues, so I will create the issues and close this topic.
